mixer/pkg/attribute/envoyProtoBag.go

```diff
@@ -0,0 +1,317 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package attribute
+
+import (
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	alspb "github.com/envoyproxy/go-control-plane/envoy/data/accesslog/v2"
+	accesslog "github.com/envoyproxy/go-control-plane/envoy/service/accesslog/v2"
+	authz "github.com/envoyproxy/go-control-plane/envoy/service/auth/v2"
+	"github.com/golang/protobuf/ptypes/timestamp"
+
+	mixerpb "istio.io/api/mixer/v1"
+	attr "istio.io/pkg/attribute"
+)
+
+// EnvoyProtoBag implements the Bag interface on top of an Attributes proto.
+type EnvoyProtoBag struct {
+	reqMap          map[string]interface{}
+	upstreamCluster string
+
+	// to keep track of attributes that are referenced
+	referencedAttrs      map[attr.Reference]attr.Presence
+	referencedAttrsMutex sync.Mutex
+}
+
+var envoyProtoBags = sync.Pool{
+	New: func() interface{} {
+		return &EnvoyProtoBag{
+			referencedAttrs: make(map[attr.Reference]attr.Presence, referencedAttrsSize),
+		}
+	},
+}
+
+func fillAddress(reqMap map[string]interface{}, address *core.Address, name string) {
+	reqMap[name+".ip"] = []byte(net.ParseIP(address.GetSocketAddress().GetAddress()).To16())
+	reqMap[name+".port"] = int64(address.GetSocketAddress().GetPortValue())
+}
+
+func reformatTime(theTime *timestamp.Timestamp, durationNanos int32) time.Time {
+	return time.Unix(theTime.Seconds, int64(theTime.Nanos)+int64(durationNanos))
+}
+
+func isGrpc(headers map[string]string) bool {
+	if val, ok := headers["content-type"]; ok {
+		return strings.Contains(val, "grpc")
```

Comment:
Starts-with 'application/grpc' might be a better approximation.

Reply:
Done